kotlinx-coroutines-core/common/src/flow/internal/Combine.kt

```diff
@@ -9,129 +9,141 @@ import kotlinx.coroutines.*
+    unsafeFlow {
+        coroutineScope {
+            val second = asChannel(flow2)
+            /*
+             * This approach only works with rendezvous channel and is required to enforce correctness
+             * in the following scenario:
+             * ```
+             * val f1 = flow { emit(1); delay(Long.MAX_VALUE) }
+             * val f2 = flowOf(1)
+             * f1.zip(f2) { ... }
+             * ```
+             *
+             * Invariant: this clause is invoked only when all elements from the channel were processed (=> rendezvous restriction).
+             */
+            val collectJob = Job()
+            val scopeJob = currentCoroutineContext()[Job]!! // TODO replace with extension when #2245 is here
+            (second as SendChannel<*>).invokeOnClose {
+                // Optimization to avoid AFE allocation when the other flow is done
+                if (collectJob.isActive) collectJob.cancel(AbortFlowException(this@unsafeFlow))
+            }
 
-        val otherIterator = second.iterator()
-        try {
-            first.consumeEach { value ->
-                if (!otherIterator.hasNext()) {
-                    return@consumeEach
+            val newContext = coroutineContext + scopeJob
+            val cnt = threadContextElements(newContext)
+            try {
+                /*
+                 * Non-trivial undispatched (because we are in the right context and there is no structured concurrency)
+                 * hierarchy:
+                 * -Outer coroutineScope that owns the whole zip process
+                 * - First flow is collected by the child of coroutineScope, collectJob.
+                 *    So it can be safely cancelled as soon as the second flow is done
+                 * - **But** the downstream MUST NOT be cancelled when the second flow is done,
+                 *    so we emit to downstream from coroutineScope job.
+                 * Typically, such hierarchy requires coroutine for collector that communicates
+                 * with coroutines scope via a channel, but it's way too expensive, so
+                 * we are using this trick instead.
+                 */
+                withContextUndispatched( coroutineContext + collectJob) {
+                    flow.collect { value ->
+                        withContextUndispatched(newContext, cnt) {
+                            val otherValue = second.receiveOrNull() ?: throw AbortFlowException(this@unsafeFlow)
+                            emit(transform(NULL.unbox(value), NULL.unbox(otherValue)))
+                        }
+                    }
                 }
-                emit(transform(NULL.unbox(value), NULL.unbox(otherIterator.next())))
+            } catch (e: AbortFlowException) {
+                e.checkOwnership(owner = this@unsafeFlow)
+            } finally {
+                if (!second.isClosedForReceive) second.cancel(AbortFlowException(this@unsafeFlow))
```

Comment:
You don't need to pass `AbortFlowException` here. `second` scope is outside this `try { ... } catch { ... }` anyway and you don't actually need to isolate second channel cancellation. Simply calling `cancel()` here works just as well.